source/slang/slang-ir-specialize-resources.cpp

```diff
@@ -263,647 +70,16 @@ struct ResourceParameterSpecializationContext
-        // to perform the second phase of cloning, which will recursively
-        // clone any nested decorations, blocks, and instructions.
-        //
-        cloneInstDecorationsAndChildren(
-            &cloneEnv,
-            builder->sharedBuilder,
-            oldFunc,
-            newFunc);
-
-        // We are almost done at this point, except that `newFunc`
-        // is lacking its parameters, as well as any of the body
-        // instructions that we decided were needed during
-        // the information-gathering steps.
-        //
-        // We will insert these instructions into the first block
-        // of the function, before its first ordinary instruction.
-        // We know that these should exist because we had as
-        // a precondition that `oldFunc` was a definition (so it
-        // has at least one block), and in valid IR every block
-        // has at least one ordinary instruction (its terminator).
-        //
-        auto newEntryBlock = newFunc->getFirstBlock();
-        SLANG_ASSERT(newEntryBlock);
-        auto newFirstOrdinary = newEntryBlock->getFirstOrdinaryInst();
-        SLANG_ASSERT(newFirstOrdinary);
-
-        // We simply iterate over the list of parameters and then
-        // body instructions that were produced in the information
-        // gathering step, and insert each before `newFirstOrdinary`,
-        // which has the effect or arranging them in the output
-        // in the order they are enumerated here.
-        //
-        for( auto newParam : funcInfo.newParams )
-        {
-            newParam->insertBefore(newFirstOrdinary);
-        }
-        for( auto newBodyInst : funcInfo.newBodyInsts )
-        {
-            newBodyInst->insertBefore(newFirstOrdinary);
-        }
-
-        // At this point we've created a new specialized function,
-        // and as such it may contain call sites that were not
-        // covered when we built our initial work list.
-        //
-        // Before handing the specialized function back to the
-        // caller, we will make sure to recursively add any
-        // potentially-specializable call sites to our work list.
-        //
-        addCallsToWorkListRec(newFunc);
-
-        return newFunc;
-    }
 };
 
-// The top-level function for invoking the specialization pass
-// is straighforward. We set up the context object
-// and then defer to it for the real work.
-//
 void specializeResourceParameters(
     BackEndCompileRequest* compileRequest,
     TargetRequest*  targetRequest,
     IRModule*       module)
 {
-    ResourceParameterSpecializationContext context;
-    context.compileRequest = compileRequest;
-    context.targetRequest = targetRequest;
-    context.module = module;
-
-    context.processModule();
+    ResourceParameterSpecializationCondition condition;
+    condition.targetRequest = targetRequest;
+    specializeFunctionCall(compileRequest, targetRequest, module, &condition);
```

Comment:
I just noticed that we should probably name the function `specializeFunctionCalls` since it can (and does) specialize multiple function calls in one go.